src/subtype.c

```diff
@@ -102,6 +99,9 @@ typedef struct jl_stenv_t {
     int intersection;         // true iff subtype is being called from intersection
     int emptiness_only;       // true iff intersection only needs to test for emptiness
     int triangular;           // when intersecting Ref{X} with Ref{<:Y}
+    // Used to represent the length difference between 2 vararg.
+    // intersect(X, Y) ==> X = Y + Loffset
+    int Loffset;
```

Comment:
I am not sure I agree with this part of the optimization in space usage. What if you have a NTuple inside a NTuple?

Reply:
As NTuple of NTuple, their `Vararg`s are not nested. 
This field would be set only when intersecting `Vararg` length.
Thus when we intersect `Tuple{S, Vararg{S, M}} where {N,S<:NTuple{N},M}`, the Loffset of N would be set only when we intersect `S`.
When we intersect `M`, the old `Loffset` should have been erased.

This rule might be broken for `Tuple{S, Vararg{S, M}} where {S, N, M<:NTuple{N}}`, but the `intvalued` check in `finish_unionall` would make sure the result is correct.
```julia
julia> typeintersect(Tuple{S, Vararg{S, M}} where {S, N, M<:NTuple{N}}, Tuple{Vararg{Int}})
Union{}
```

Reply:
Thanks for explaining! I was also assuming it might also get used for a case like this, where we we have to set M=N+1 and N=M+1, then leading to failure. But apparently we just return a conservative answer:
```
julia> typeintersect(Tuple{Tuple{Int,Vararg{Int,N}},Tuple{Vararg{Int,N}}} where N,
                     Tuple{Tuple{Vararg{Int,M}},Tuple{Int,Vararg{Int,M}}} where M)
Tuple{Tuple{Int64, Vararg{Int64, N}}, Tuple{Int64, Vararg{Int64, M}}} where {N, M}
```

Reply:
Oh yes. This is similar to #39099, the offset between 2 typevars is ignored during subtyping/intersection.
https://github.com/JuliaLang/julia/pull/48555 aims to fix this, but it breaks the the reliability of `==` check.
And it doesn't play well with the var chain during subtyping. (e,g, `NTuple{N,Any} where {M,N<:M}`)